daemon/graphdriver/devmapper/deviceset.go

```diff
@@ -553,6 +580,62 @@ func (devices *DeviceSet) migrateOldMetaData() error {
 	return nil
 }
 
+// Cleanup deleted devices. It assumes that all the devices have been
+// loaded in the hash table. Should be called with devices.Lock() held.
+// Will drop the lock for device deletion and return with lock acquired.
+func (devices *DeviceSet) cleanupDeletedDevices() error {
+	var deletedDevices []*devInfo
+
+	// If there are no deleted devices, there is nothing to do.
+	if devices.nrDeletedDevices == 0 {
+		return nil
+	}
+
+	for _, info := range devices.Devices {
```

Comment:
btw you have same loop here and in countDeletedDevices, but you use different rule for condition(continue on false vs action on true), I'm not against any, but it should be same I think.